browser/brave_wallet/ethereum_provider_impl_unittest.cc

```diff
@@ -1964,6 +1986,62 @@ TEST_F(EthereumProviderImplUnitTest, AccountsChangedEvent) {
   EXPECT_FALSE(observer_->AccountsChangedFired());
 }
 
+TEST_F(EthereumProviderImplUnitTest, EthSubscribe) {
+  CreateWallet();
+
+  // Unsupported subscription type
+  std::string request_payload_json =
+      R"({"id":1,"jsonrpc:": "2.0","method":"eth_subscribe",
+          "params": ["foo"]})";
+  absl::optional<base::Value> request_payload = base::JSONReader::Read(
+      request_payload_json, base::JSON_PARSE_CHROMIUM_EXTENSIONS |
+                                base::JSONParserOptions::JSON_PARSE_RFC);
+  auto response = CommonRequestOrSendAsync(request_payload.value());
+
+  mojom::ProviderError error_code;
+  std::string error_message;
+  GetErrorCodeMessage(std::move(response.second), &error_code, &error_message);
+  EXPECT_EQ(response.first, true);
+  EXPECT_EQ(error_code, mojom::ProviderError::kInternalError);
+  EXPECT_EQ(error_message,
+            l10n_util::GetStringUTF8(IDS_WALLET_UNSUPPORTED_SUBSCRIPTION_TYPE));
+
+  url_loader_factory_.SetInterceptor(
+      base::BindLambdaForTesting([&](const network::ResourceRequest& request) {
+        url_loader_factory_.ClearResponses();
+
+        std::string header_value;
+        EXPECT_TRUE(request.headers.GetHeader("X-Eth-Method", &header_value));
+        std::string content;
+        if (header_value == "eth_blockNumber" ||
+            header_value == "getBlockHeight")
+          content = R"({"id":1,"jsonrpc":"2.0","result":"0x131131"})";
+        else if (header_value == "eth_getBlockByNumber")
+          content = R"({"id":1,"jsonrpc":"2.0","result":{"difficulty":"0x1"}})";
+        url_loader_factory_.AddResponse(request.url.spec(), content);
+      }));
+
+  // Subscribing to newHeads
+  request_payload_json =
+      R"({"id":1,"jsonrpc:": "2.0","method":"eth_subscribe",
+          "params": ["newHeads"]})";
+  request_payload = base::JSONReader::Read(
+      request_payload_json, base::JSON_PARSE_CHROMIUM_EXTENSIONS |
+                                base::JSONParserOptions::JSON_PARSE_RFC);
+  response = CommonRequestOrSendAsync(request_payload.value());
+  EXPECT_EQ(response.first, false);
+  EXPECT_TRUE(response.second.is_string());
+  browser_task_environment_.FastForwardBy(
+      base::Seconds(kBlockTrackerDefaultTimeInSeconds));
+  EXPECT_TRUE(observer_->MessageEventFired());
+  base::Value rv = observer_->GetLastMessage();
+  ASSERT_TRUE(rv.is_dict());
+  auto& dict = rv.GetDict();
+  std::string* difficulty = dict.FindString("difficulty");
+  ASSERT_TRUE(difficulty);
+  EXPECT_EQ(*difficulty, "0x1");
```

Comment:
Can we also do `eth_unsubscribe` with non existent subscription id and check `EthBlockTracker` is still running and the last subscription id to check if `EthBlockTracker` stop running?

Reply:
Fixed in 60e51cb1134d0b20bb67da65ffdbd2808d25b0be
I also added a test to make sure it doesn't stop when a 2nd subscription exists.